File: src/formats/ernest.ts
import {ReactionNetwork, Reaction, Complex} from '../reactions';

export function generateFile(rn: ReactionNetwork): string {
  let outputLines = [];
  function addLine(...args) {
    outputLines.push(fillBlanks.apply(null, args));
  }

  addLine("clear model");
  addLine("");
  
  let name = rn.modelName;
  addLine("model.id = '__';", name.replace(/\s+/g, '_'));
  addLine("model.name = '__';", name);
  
  let speciesList = rn.getSpecies().sort().map(enquote);
  addLine("model.species = struct('id', __);", speciesList);
  let reactions = rn.reactions;
  reactions.forEach((r, i) =>
    addLine("model.reaction(__) = __;", (i+1), convertReaction(r))
  );
  
  addLine("");
  addLine("disp(['Model: ' model.name]);");
  addLine("model_analysis(model);");
  addLine("");
  
  return outputLines.join("\n");
}

function convertReaction(r: Reaction): string {
  let id = reactionString(r);
  let reactant = convertComplex(r.reactant);
  let product = convertComplex(r.product);
  return fillBlanks(
    "struct('id', __, 'reactant', __, 'product', __, 'reversible', __)",
      enquote(id), reactant, product, r.reversible
  );
}

function reactionString(r: Reaction): string {
  let left = r.reactant.join("+") || "0";
  let right = r.product.join("+") || "0";
  let arrow = r.reversible ? "<->" : "->";
  return left + arrow + right;
}

function convertComplex(complex: Complex): string {
  if (complex.length === 0) {
    return "struct([])";
  }

  let species = complex.map(term => enquote(term.species));
  let stoich = complex.map(term => term.coefficient);
  return fillBlanks("struct('species', __, 'stoichiometry', __)", species, stoich);
}

function fillBlanks(template: string, ...replacements): string {
  let i = 0;
  return template.replace(/__/g, () => {
    let r = replacements[i++];
    return r instanceof Array ? encodeArray(r) : r;
  });
}

function enquote(str: string): string {
  return "'" + str + "'";
}

function encodeArray(arr: Array<any>): string {
  if (arr.length === 1) {
      return arr[0];
  } else {
      return "{" + arr.join(", ") + "}";
  }
}

File: samples/ERNEST.m
clear model

model.id = 'Sample_Model';
model.name = 'Sample Model';
model.species = struct('id', {'A', 'A1', 'C4', 'NC', 'S', 'S7' 'X1', 'X2', 'X3', 'X5'});
model.reaction(1) = struct('id', 'X1<->7X2', 'reactant', struct('species', 'X1', 'stoichiometry', 1), 'product', struct('species', 'X2', 'stoichiometry', 7), 'reversible', true);
model.reaction(2) = struct('id', '2X3+X2<-C4', 'reactant', struct('species', 'C4', 'stoichiometry', 1), 'product', struct('species', {'X3', 'X2'}, 'stoichiometry', {2, 1}), 'reversible', false);
model.reaction(3) = struct('id', 'A+NC+7S->A1+S7', 'reactant', struct('species', {'A', 'NC', 'S'}, 'stoichiometry', {1, 1, 7}), 'product', struct('species', {'A1', 'S7'}, 'stoichiometry', {1, 1}), 'reversible', false);
model.reaction(4) = struct('id', '0->X5', 'reactant', struct([]), 'product', struct('species', 'X5', 'stoichiometry', 1), 'reversible', false);

disp(['Model: ' model.name]);
model_analysis(model);
